### PR TITLE
add support for parsing bitbucket urls

### PR DESCRIPTION
**Pull request by matthyx**

### CODE REVIEW

File: apis/bitbucketapi/apis.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+type BitBucketAPI struct {
+	httpClient *http.Client
+}
+
+func NewBitBucketAPI() *BitBucketAPI { return &BitBucketAPI{httpClient: &http.Client{}} }
+
+func (gl *BitBucketAPI) GetRepoTree(owner, repo, branch string, headers *Headers) (*Tree, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (dwertent) [inline apis/bitbucketapi/apis.go]:
```suggestion
        return nil, fmt.Errorf("GetRepoTree is not supported")
```
This function should return an error indicating that it is not implemented, but not panic.

File: apis/bitbucketapi/apis.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func (gl *BitBucketAPI) GetDefaultBranchName(owner, repo string, headers *Headers) (string, error) {
+	id := owner + "%2F" + repo
```
Comment (dwertent) [inline apis/bitbucketapi/apis.go]:
```suggestion
	id := url.Join(owner, repo)
```

File: apis/bitbucketapi/apis.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	var data bitbucketBranchingModel
+	if err = json.Unmarshal([]byte(body), &data); err != nil {
+		return "", err
+	}
+
+	return data.Development.Name, nil
+}
+
+func (gl *BitBucketAPI) GetLatestCommit(owner, repo, branch string, headers *Headers) (*Commit, error) {
+	id := owner + "%2F" + repo
```
Comment (dwertent) [inline apis/bitbucketapi/apis.go]:
```suggestion
	id := url.Join(owner, repo)
```

File: apis/bitbucketapi/apis.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		return nil, err
+	}
+	return &data.Values[0], nil
+}
+
+// APIBranchingModel
+// API Ref: https://developer.atlassian.com/cloud/bitbucket/rest/api-group-branching-model/#api-group-branching-model
+// Example: https://bitbucket.org/!api/2.0/repositories/matthyx/ks-testing-public/branching-model
+func APIBranchingModel(id string) string {
+	return fmt.Sprintf("https://%s/!api/2.0/repositories/%s/branching-model", DEFAULT_HOST, id)
```
Comment (dwertent) [inline apis/bitbucketapi/apis.go]: It's nicer to use the `url` package, but it is not a must

File: bitbucketparser/v1/commit.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	return bitBucketAPICommitToCommit(c), nil
+}
+
+func bitBucketAPICommitToCommit(c *bitbucketapi.Commit) *apis.Commit {
+	name, email := parseRawUser(c.Author.Raw)
+	latestCommit := &apis.Commit{
+		SHA: c.Hash,
+		Author: apis.Committer{
```
Comment (dwertent) [inline bitbucketparser/v1/commit.go:36]: We should add a comment explaining that the API does not return the `Author`

File: bitbucketparser/v1/tree.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package v1
+
+func (gl *BitBucketURL) ListAllNames() ([]string, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (dwertent) [inline bitbucketparser/v1/tree.go]:
```suggestion
	return []string{}, fmt.Errorf("ListAllNames is not supported")
```

File: bitbucketparser/v1/tree.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+func (gl *BitBucketURL) ListDirsNames() ([]string, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (dwertent) [inline bitbucketparser/v1/tree.go]:
```suggestion
	return []string{}, fmt.Errorf("ListDirsNames is not supported")
```

File: bitbucketparser/v1/tree.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+func (gl *BitBucketURL) ListFilesNames() ([]string, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (dwertent) [inline bitbucketparser/v1/tree.go]:
```suggestion
	return []string{}, fmt.Errorf("ListFilesNames is not supported")
```

File: bitbucketparser/v1/tree.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+func (gl *BitBucketURL) ListFilesNamesWithExtension(extensions []string) ([]string, error) {
+	//TODO implement me
+	panic("implement me")
```
Comment (dwertent) [inline bitbucketparser/v1/tree.go]:
```suggestion
	return []string{}, fmt.Errorf("ListFilesNamesWithExtension is not supported")
```

**Review by dwertent — CHANGES_REQUESTED** [review]

File: apis/bitbucketapi/apis.go (hunk 4 above)
Comment (matthyx) [reply apis/bitbucketapi/apis.go]: I see the need to escape variables from URL, so something like that instead (example taken from somewhere else to show a query)?
```
	u := url.URL{
		Scheme: "https",
		Host:   DEFAULT_HOST,
		Path:   fmt.Sprintf("api/v4/projects/%s/repository/tree", id),
		RawQuery: fmt.Sprintf("ref=%s", branch),
	}
	return u.String()
```

Comment (dwertent) [reply apis/bitbucketapi/apis.go]: For the path, you should use the `url.Join` method

**Review by dwertent — APPROVED** [review]

**Pull request merged**